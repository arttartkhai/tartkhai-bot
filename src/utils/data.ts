import { ILatestQuote, IMetadata } from 'types/services'
import {CustomCoinData } from 'types/template'

const formatNumber = (num: number): string => {
  return num.toLocaleString(undefined, {maximumFractionDigits: 2})
}

export const createCustomPriceData = (meta: IMetadata, quote: ILatestQuote) : CustomCoinData => ({
  name: meta.name,
  symbol: meta.symbol,
  logo: meta.logo,
  maxSupply: quote.max_supply ? formatNumber(quote.max_supply) : null,
  totalSupply: formatNumber(quote.total_supply),
  rank: formatNumber(quote.cmc_rank),
  price: formatNumber(quote.quote?.USD?.price),
  volume: formatNumber(quote.quote?.USD?.volume_24h),
  marketCap: formatNumber(quote.quote?.USD?.market_cap),
  percentChange_1h: formatNumber(quote.quote?.USD?.percent_change_1h),
  percentChange_24h: formatNumber(quote.quote?.USD?.percent_change_24h),
  percentChange_7d: formatNumber(quote.quote?.USD?.percent_change_7d),
  percentChange_30d: formatNumber(quote.quote?.USD?.percent_change_30d),
  percentChange_60d: formatNumber(quote.quote?.USD?.percent_change_60d),
  percentChange_90d: formatNumber(quote.quote?.USD?.percent_change_90d),
})
