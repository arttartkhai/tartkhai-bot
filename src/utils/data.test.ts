import {createCustomPriceData} from './data'

describe('createCustomPriceData', () => {
    const mockMeta = {
        "id": 1,
        "name": "Bitcoin",
        "symbol": "BTC",
        "category": "coin",
        "description": "Bitcoin (BTC) is a cryptocurrency . Users are able to generate BTC through the process of mining. Bitcoin has a current supply of 18,783,500. The last known price of Bitcoin is 45,616.08914537 USD and is down -0.12 over the last 24 hours. It is currently trading on 8933 active market(s) with $34,409,712,312.26 traded over the last 24 hours. More information can be found at https://bitcoin.org/.",
        "slug": "bitcoin",
        "logo": "https://s2.coinmarketcap.com/static/img/coins/64x64/1.png",
        "urls": {
            "website": [
                "https://bitcoin.org/"
            ],
        },
        "platform": null,
    }
    const mockPrice = {
        "num_market_pairs": 8938,
        "max_supply": 21000000,
        "circulating_supply": 18779718,
        "total_supply": 18779718,
        "cmc_rank": 1,
        "quote": {
            "USD": {
                "price": 45049.61289091343,
                "volume_24h": 39707372463.017365,
                "percent_change_1h": 0.69419495,
                "percent_change_24h": 4.12124901,
                "percent_change_7d": 7.9247753,
                "percent_change_30d": 37.09556361,
                "percent_change_60d": 31.55031237,
                "percent_change_90d": -23.46503889,
                "market_cap": 846019026100.5189,
            }
        }
    }
    
    it('should return correct custom data', () => {
        expect(createCustomPriceData(mockMeta, mockPrice)).toMatchSnapshot()
    })
})