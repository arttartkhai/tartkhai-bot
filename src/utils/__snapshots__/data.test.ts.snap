// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createCustomPriceData should return correct custom data 1`] = `
Object {
  "logo": "https://s2.coinmarketcap.com/static/img/coins/64x64/1.png",
  "marketCap": "846019026100.52",
  "maxSupply": 21000000,
  "name": "Bitcoin",
  "percentChange_1h": "0.69",
  "percentChange_24h": "4.12",
  "percentChange_30d": "37.10",
  "percentChange_60d": "31.55",
  "percentChange_7d": "7.92",
  "percentChange_90d": "-23.47",
  "price": "45049.61",
  "rank": 1,
  "symbol": "BTC",
  "totalSupply": 18779718,
  "volume": "39707372463.02",
}
`;
